### PR TITLE
feat(db): an atomic multi-statement transaction on the shared connection

### PR DESCRIPTION
**Pull request by WingedGuardian**
Split out of #1576 so it can be reviewed as what it actually is: a change to the connection layer **every subsystem's database access goes through**, rather than an appendix to a dark subsystem. #1576 keeps the federation code and rebases onto this once it lands.

## Why split

#1576 is 3,486 lines across 25 files and has been through five cross-model rounds against a cap of three. Five of its findings open with the words *"Fresh evidence after the earlier fix"* — the reviewer naming the churn. It bundles ~2,900 lines of a dark subsystem, whose consumer is unwritten, with a rewrite of the shared connection layer. All the live risk is in the second part, and reviewing an interface with no consumer is why each round kept finding new things.

The split paid for itself immediately. One focused audit of this code alone surfaced a defect five rounds on the combined PR had not — see below.

## What it does

Each statement on the shared connection is individually serialised; a *sequence* of them is not. A compare-and-set UPDATE followed by the INSERT that depends on it can be split apart by another coroutine committing in between. `transaction()` holds the lock across `BEGIN IMMEDIATE` … COMMIT, with statements from the same task running on the already-held lock. COMMIT on clean exit, ROLLBACK on any exception, then re-raise.

## The refusal surface, which is where the care went

A partial guard on an atomicity contract is worse than none, and this one was partial in three ways:

- **Transaction-control SQL was refused on only two of the four SQL-taking methods.** `execute_fetchall` and `execute_insert` forwarded it.
- **Guarding `cursor()` was theatre.** `execute()` returns an `aiosqlite.Cursor` too, so the same class was reachable through the ordinary query path. Measured, three escapes: `await cur.execute("COMMIT")`, the same through `async with`, and `cur.executescript(...)` — each ended the transaction while the context manager still reported success. The refusal now lives on a wrapper applied at every cursor-returning site, and is a no-op outside a transaction so the hot path is unchanged.
- **A UTF-8 BOM walked past the verb parser.** Python's `\s` doesn't match U+FEFF; SQLite skips it. So a BOM-prefixed COMMIT read as verbless and executed for real. Reachable from any `.sql` file saved by a BOM-writing editor, since reading with plain UTF-8 keeps a BOM.

## Risk today is nil, which is the argument for landing it now

`transaction()` has no callers — the consumer stays with the federation PR. So the ownership field is permanently unset in production, the conditional lock reduces to the previous unconditional one, and the five rewritten hot-path methods are behaviour-identical to `main` until the first caller lands. Every risk here is latent, and this is the cheapest moment to close it.

Enumerated rather than assumed: of the twelve transaction-control call sites across `src/` and `scripts/`, exactly one family routes through this class (the migration runner). `cursor()` has zero production callers. A docstring claiming the precompact script as a caller is corrected — it holds a raw stdlib handle and never reaches this class.

## Open decisions I did not make unilaterally

Three findings are named here rather than silently carried, because two are design calls and one widens the diff:

1. **An unbounded transaction body holds the process-wide connection lock**, and the entry-wait wedge detector structurally cannot diagnose it — it runs only after acquiring the lock being held. A whole-process stall was measured, with the reconnect recovery path deliberately disabled during a transaction. The obvious remedy is a default body timeout, but that is a real design choice about a primitive, with its own failure mode: killing a legitimately slow transaction. Worth deciding before the first caller.
2. **The verb parser is duplicated in the migration runner, and the copy there handles no comments**, so a commented transaction-control statement is refused here and accepted in a migration — the weaker parser guarding the higher-stakes path. The fix is for the runner to import from this module (this file is the lower level), which widens the diff into the migration runner, so it is raised rather than bundled into a split whose point was to narrow the surface.
3. **A child task touching the connection inside a block deadlocks unbreakably** if the parent awaits it (a `gather` inside a body). The docstring says "BLOCKS", which reads as "resolves at commit". Wording and a test belong with decision 1, since a body timeout converts the deadlock into a raised error.

## Verification

82 tests pass across the connection, write-retry and migration-runner suites. Both new guards verify-RED against a mutant with them removed, each failing on the primary assertion. The enumerating test asserts an exact set rather than a floor, so losing a method fails as loudly as adding one. One internal adversarial audit; every finding either fixed or named above.

E2E: after merge, no runtime behaviour changes until a caller exists — `transaction()` has no call sites, so the deployed path is provably identical. The E2E that matters belongs to the first consumer (#1576): a concurrent read-modify-write under load must apply both statements or neither.

---

<a href="https://app.devin.ai/review/wingedguardian/genesis-agi/pull/1881" target="_blank"><picture><source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=4">Devin Review</picture></a>

## Summary by CodeRabbit

- **New Features**
  - Added atomic database transactions with commit and rollback support.
  - Prevented transaction-control commands from bypassing managed transaction ownership.
  - Protected database cursors during active transactions.
  - Added safeguards for nested transactions and concurrent database access.

- **Bug Fixes**
  - Improved handling of contention, cancellations, lock errors, and interrupted transactions.
  - Added recovery and reconnection safeguards when transactions cannot be rolled back safely.
  - Improved isolation and reliability for concurrent database operations.
  - Ensured connections are safely quarantined when transaction protection cannot be restored.

### CODE REVIEW

**Comment by WingedGuardian** [comment]
## Automated first-pass review (structural + PII/secrets/wording)

**Scope:** 3 files changed — `changelog.d/20260909020000-added-atomic-db-transaction-primitive.md`, `src/genesis/db/connection.py`, `tests/test_db/test_connection.py`. Not documentation-only, so both passes ran in full.

### Domain 1 — Structural review

Checked against the pattern-violation, import-hygiene, coverage, documentation, and Genesis-convention checklists.

- No `nohup`, no direct `task_states` inserts, no `/tmp/` usage, no `git add -A`/`--force`/`--hard`, no worktree-path `pip install -e`, no bare `except:`, no mutable default arguments, no `os.killpg` misuse.
- Raw SQL literals (`BEGIN IMMEDIATE`, `COMMIT`, `ROLLBACK`) appear only inside `src/genesis/db/connection.py` itself — this *is* the low-level connection primitive underlying `genesis.db.crud`, not application code bypassing it, so the "raw SQL against genesis.db" rule doesn't apply here.
- Import hygiene: `import re` is stdlib and correctly alphabetized among the other stdlib imports (asyncio, logging, random, re, sqlite3); no circular or star imports.
- `_GuardedCursor` references `SerializedConnection` as a type hint before the latter is defined in the file — not a bug, since `from __future__ import annotations` (line 8) makes this a lazily-evaluated string annotation.
- Documentation: every new class/function (`_GuardedCursor`, `_leading_sql_verb`, `transaction()`, and the private `_owns_txn`/`_maybe_lock`/`_guard_cursor`/`_refuse_inside_txn`/`_refuse_txn_control_sql` helpers) carries a docstring explaining rationale, not just behavior.
- Coverage: the diff adds ~155 lines of production code and ~398 lines of new tests in the same PR covering commit/rollback semantics, lock re-entrancy, peer isolation, non-reentrancy, the implicit-transaction wait (wedge vs. contention), transaction-control-SQL refusal on all four SQL-taking methods (asserted as an exact set, not a floor), the cursor-escape hole, and the BOM-verb-parsing edge case. No gap identified.
- Commit/PR title carries a conventional prefix (`feat(db): ...`).

No structural issues found in this pass.

### Domain 2 — PII, Secrets, and Wording

- Added lines (code + tests): no API keys/tokens/passwords, no private IPv4/IPv6 ranges, no SSH keys, no DB connection strings with credentials, no secret env values.
- No personal emails, no usernamed absolute paths, no location-revealing timezone identifiers, no cloud account/project IDs, no hardware identifiers, no private hostnames.
- PR title/body/changelog: describes a database-concurrency defect (transaction atomicity, cursor-guard gap, BOM parsing edge case) in generic technical terms. No reference to a live security incident, no naming of private infrastructure, and no exploit-oriented detail beyond what's needed to justify the fix in a public engineering repo.

PII/Secrets/Wording: clean (no soft warnings either).

### Verdict

Structural review: no pattern violations found.
PII/Secrets/Wording scan: clean.

VERDICT: PASS




---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/genesis/db/connection.py [hunk 1]
```diff
@@ -103,6 +139,61 @@ def _is_lock_error(exc: BaseException) -> bool:
...
+    def __getattr__(self, name: str) -> Any:
+        return getattr(object.__getattribute__(self, "_cur"), name)
+
+    async def execute(self, sql: str, parameters: Iterable[Any] | None = None) -> Any:
+        self._owner._refuse_txn_control_sql(sql)
+        return await self._cur.execute(sql, parameters)
+
+    async def executemany(self, sql: str, parameters: Iterable[Iterable[Any]]) -> Any:
+        self._owner._refuse_txn_control_sql(sql)
+        return await self._cur.executemany(sql, parameters)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/db/connection.py]: Fixed at 6c7d6ffa2: _GuardedCursor.execute/executemany now return the wrapper, so the chained form cur = await cur.execute(...) keeps the guard on every hop — a chained execute("COMMIT") inside transaction() now hits the refusal instead of bypassing it. Regression-pinned in tests/test_db/test_connection.py.

File: src/genesis/db/connection.py [hunk 2]
```diff
@@ -346,12 +548,186 @@ async def rollback(self) -> None:
...
+            object.__setattr__(self, "_txn_owner", asyncio.current_task())
+            began = False
+            try:
+                # Explicit IMMEDIATE: takes the write lock up front (no lazy
+                # read→write upgrade deadlock across processes) and makes the
+                # boundary unambiguous rather than relying on legacy implicit-BEGIN
+                # timing. A locked BEGIN opened no transaction, so _retry_locked
+                # re-running it is safe. Matches the migration runner's idiom.
+                await self._retry_locked(lambda: self._conn.execute("BEGIN IMMEDIATE"))
+                began = True
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/db/connection.py]: Fixed at 6c7d6ffa2: boundary outcomes are now resolved via _txn_is_open()'s settle probe (a serialized SELECT 1 drains aiosqlite's worker queue before in_transaction is read). A cancelled BEGIN await lands in the except path, rolls back a provably-open transaction, and quarantines; ownership is claimed only after BEGIN lands. Main merged in at 0902afdb0 (union with the corrupt-DB quarantine) — suite green.

File: src/genesis/db/connection.py [hunk 3]
```diff
@@ -346,12 +548,186 @@ async def rollback(self) -> None:
...
+                await self._retry_locked(lambda: self._conn.commit())
+            except BaseException:
+                if began:
+                    try:
+                        await self._retry_locked(lambda: self._conn.rollback())
+                    except Exception:
+                        # A failed ROLLBACK must not mask the original error. The
+                        # connection may wedge in_transaction=True until the
+                        # reconnect threshold recovers it; log and re-raise the
+                        # real cause.
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/db/connection.py]: Fixed at 6c7d6ffa2: _rollback_then_quarantine closes and reconnects when a rollback fails with the transaction possibly open, so a later caller can never commit the half-applied unit; a CancelledError during the rollback is state-checked rather than used to skip quarantine.

File: src/genesis/db/connection.py [hunk 4]
```diff
@@ -176,7 +282,14 @@ async def _handle_lock_error(self, exc: Exception) -> None:
...
         object.__setattr__(self, "_consecutive_errors", count)
-        if count >= self._max_errors and self._reconnect_fn is not None:
+        # Never reconnect in the middle of an OPEN explicit transaction: swapping
+        # `_conn` here would silently discard the uncommitted transaction on the
+        # old connection and leave :meth:`transaction`'s COMMIT/ROLLBACK operating
+        # on a fresh, transaction-less connection. Integrity is preserved either
+        # way (the CAS + single commit means a discarded txn wrote nothing), so
+        # just count the episode and raise — transaction() then rolls back on the
+        # still-open connection and the caller retries the whole unit. (F-E)
+        if count >= self._max_errors and self._reconnect_fn is not None and not self._owns_txn():
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/db/connection.py:532]: Fixed at 6c7d6ffa2: _txn_owner is set only AFTER BEGIN succeeds, so a pre-BEGIN state no longer reads as an open transaction — a repeatedly failing BEGIN episode reaches _handle_lock_error's reconnect path and the connection recovers.

**Comment by WingedGuardian** [comment]
@codex review
